Shared/Realm.Sync.Shared/User.cs

```diff
@@ -127,25 +128,53 @@ public void LogOut()
                 request.Headers.Accept.ParseAdd(_applicationProblemJsonUtf8MediaType.MediaType);
 
                 var response = await client.SendAsync(request).ConfigureAwait(continueOnCapturedContext: false);
-                if (response.StatusCode == HttpStatusCode.OK && response.Content.Headers.ContentType.Equals(_applicationJsonUtf8MediaType))
+                if (response.IsSuccessStatusCode && response.Content.Headers.ContentType.Equals(_applicationJsonUtf8MediaType))
                 {
                     using (var stream = await response.Content.ReadAsStreamAsync().ConfigureAwait(continueOnCapturedContext: false))
                     {
                         return JsonValue.Load(stream);
                     }
                 }
-                else if (response.Content.Headers.ContentType.Equals(_applicationProblemJsonUtf8MediaType))
+
+                using (var stream = await response.Content.ReadAsStreamAsync().ConfigureAwait(continueOnCapturedContext: false))
                 {
-                    using (var stream = await response.Content.ReadAsStreamAsync().ConfigureAwait(continueOnCapturedContext: false))
+                    if (response.Content.Headers.ContentType.Equals(_applicationProblemJsonUtf8MediaType))
                     {
                         var problem = JsonValue.Load(stream);
-                        throw new Exception("AuthError"); // FIXME: parse the error details and throw a more specific exception
+
+                        ErrorCode code;
+                        if (!Enum.TryParse(problem["code"], out code))
+                        {
+                            code = ErrorCode.Unknown;
+                        }
+
+                        throw new AuthenticationException(code, response.StatusCode, response.ReasonPhrase, problem.ToString(), problem["title"]);
```

Comment:
Would it make sense to expose the problem payload on the `AuthenticationException` as a `JsonObject` property so that provider-specific details can be accessed programatically by the developer?